### ISSUE TITLE
Cache or pass only relevant information to routes

### ISSUE THREAD

**Issue by jelly** [jelly/plugin.video.fosdem#4]
A kodi addon runs every time there is UI interaction and will fetch the whole XML file every time.

* Either cache the XML file if that's possible from a Kodi addon
* Parse all data upfront and pass relevant data to views
* Keep it as is.

**Comment by yoe** [comment]
Is there anything we, the FOSDEM video team, could do to make that easier? It does seem to take a while on my kodi installation to download things ;-)

**Comment by jelly** [comment]
@yoe currently the addon downloads the schedule in [XML](https://fosdem.org/2019/schedule/xml) which is 1.6 MB and parses it for video urls. Downloading and parsing the XML file is certainly non-optional. If a lighter (json) endpoint with events and video urls is available that would be awesome.

**Comment by jelly** [comment]
@yoe actually it's elementree which parses xml which is slow:
```
2019-08-30 18:12:31.256 T:1564472176   DEBUG: show_dir
2019-08-30 18:12:31.256 T:1564472176   DEBUG: start etch_xml
2019-08-30 18:12:31.256 T:1564472176   DEBUG: start request
2019-08-30 18:12:31.784 T:1564472176   DEBUG: end request, start parsing
2019-08-30 18:13:38.292 T:1564472176   DEBUG: end parsing
2019-08-30 18:13:38.293 T:1564472176   DEBUG: end fetch_xml
2019-08-30 18:13:38.293 T:1564472176   DEBUG: start for
2019-08-30 18:13:38.294 T:1564472176   DEBUG: end for
2019-08-30 18:13:38.294 T:1564472176   DEBUG: end show_dir
```

**Comment by jelly** [comment]
Further debugging found that the problem is getting r.text from requests which takes 1 minute. Simply using urllib takes 1 second.

**Comment by dagwieers** [comment]
Yes, requests is unusable on devices like Raspberry Pi since importing urllib3 causes massive regular expressions to be compiled. We recommend all Kodi add-on developers to ditch requests/urllib3 in favor of urllib/urllib2 (python3/python2).

**Comment by dagwieers** [comment]
It is also recommended to use `reuselanguageinvoker` in your addon.xml to speed up your addon.

**Comment by jelly** [comment]
I've tried using the reuselanguageinvoker without caching for now and it didn't make it much faster still ~ 2 seconds.

It does require more changes.
```
+import sys
 from resources.lib.main import *

 try:
@@ -373,4 +374,4 @@ def iagl_text_viewer():
        IAGL_Dialog.doModal()

 if __name__ == '__main__':
-       plugin.run()
+       plugin.run(sys.argv)
```

**Comment by jelly** [comment]
Further testing on a RPI3 I figured out that urllib2 versus requests isn't a changing a lot. Actually the main issue seems to be parsing XML:

```
2019-08-30 20:11:11.201 T:1593222000   DEBUG: start request
2019-08-30 20:11:11.267 T:1937418640   DEBUG: ------ Window Init (DialogBusy.xml) ------
2019-08-30 20:11:11.996 T:1593222000   DEBUG: end request
2019-08-30 20:11:11.997 T:1593222000   DEBUG: start parsing
2019-08-30 20:11:13.386 T:1593222000   DEBUG: end parsing
```

**Comment by dagwieers** [comment]
You have to store the processed data, and only reprocess it if it wasn't stored.
That should save you the reprocessing of the XML on every action.
(But this is only possible with reuselanguageinvoker)

The requests/urllib3 impact is a 1 second hit on import now, if you use reuselanguageinvoker, it is only a one time 1 second hit if you stay within the add-on.

**Comment by jelly** [comment]
Sure I understand the concept of caching but whatever I can fix by doing smarter things upfront is what I prefer :) Sadly Kodi does not support C extensions

**Comment by yoe** [comment]
> @yoe currently the addon downloads the schedule in [XML](https://fosdem.org/2019/schedule/xml) which is 1.6 MB and parses it for video urls. Downloading and parsing the XML file is certainly non-optional. If a lighter (json) endpoint with events and video urls is available that would be awesome.

We've been working on a JSON version of the video metadata, actually; it's just not (yet) available for all years, because while there is a tool to convert, it's not entirely bugfree (and so it needs some handholding).

You can find the repository at https://github.com/FOSDEM/video-meta (and whoops, I still need to commit the 2019 data to it).

**Comment by dagwieers** [comment]
> Sure I understand the concept of caching but whatever I can fix by doing smarter things upfront is what I prefer :)

This is exactly what I meant :+1:  https://github.com/jelly/plugin.video.fosdem/commit/452a7bc9c65a22b140349103421e0de471dd03f6
